packages/core/src/BinaryData/BinaryData.service.ts

```diff
@@ -0,0 +1,217 @@
+
+			await this.managers.filesystem.init(mainManager);
+		}
+	}
+
+	async copyBinaryFile(binaryData: IBinaryData, path: string, executionId: string) {
+		const manager = this.managers[this.mode];
+
+		if (!manager) {
+			const { size } = await stat(path);
+			binaryData.fileSize = prettyBytes(size);
+			binaryData.data = await readFile(path, { encoding: BINARY_ENCODING });
+
+			return binaryData;
+		}
+
+		const identifier = await manager.copyByPath(path, executionId);
+		binaryData.id = this.createIdentifier(identifier);
+		binaryData.data = this.mode; // clear binary data from memory
+
+		const fileSize = await manager.getSize(identifier);
+		binaryData.fileSize = prettyBytes(fileSize);
+
+		await manager.storeMetadata(identifier, {
+			fileName: binaryData.fileName,
+			mimeType: binaryData.mimeType,
+			fileSize,
+		});
+
+		return binaryData;
+	}
+
+	async store(binaryData: IBinaryData, input: Buffer | Readable, executionId: string) {
+		const manager = this.managers[this.mode];
+
+		if (!manager) {
+			const buffer = await this.binaryToBuffer(input);
+			binaryData.data = buffer.toString(BINARY_ENCODING);
+			binaryData.fileSize = prettyBytes(buffer.length);
+
+			return binaryData;
+		}
+
+		const identifier = await manager.store(input, executionId);
+		binaryData.id = this.createIdentifier(identifier);
+		binaryData.data = this.mode; // clear binary data from memory
+
+		const fileSize = await manager.getSize(identifier);
+		binaryData.fileSize = prettyBytes(fileSize);
+
+		await manager.storeMetadata(identifier, {
+			fileName: binaryData.fileName,
+			mimeType: binaryData.mimeType,
+			fileSize,
+		});
+
+		return binaryData;
+	}
+
+	async binaryToBuffer(body: Buffer | Readable) {
+		return new Promise<Buffer>((resolve) => {
+			if (Buffer.isBuffer(body)) resolve(body);
+			else body.pipe(concatStream(resolve));
+		});
+	}
+
+	getAsStream(identifier: string, chunkSize?: number) {
+		const { mode, id } = this.splitBinaryModeFileId(identifier);
+
+		return this.getManager(mode).getStream(id, chunkSize);
+	}
+
+	async getBinaryDataBuffer(binaryData: IBinaryData) {
+		if (binaryData.id) return this.retrieveBinaryDataByIdentifier(binaryData.id);
+
+		return Buffer.from(binaryData.data, BINARY_ENCODING);
+	}
+
+	async retrieveBinaryDataByIdentifier(identifier: string) {
+		const { mode, id } = this.splitBinaryModeFileId(identifier);
+
+		return this.getManager(mode).getBuffer(id);
+	}
+
+	getPath(identifier: string) {
+		const { mode, id } = this.splitBinaryModeFileId(identifier);
+
+		return this.getManager(mode).getPath(id);
+	}
+
+	async getMetadata(identifier: string) {
+		const { mode, id } = this.splitBinaryModeFileId(identifier);
+
+		return this.getManager(mode).getMetadata(id);
+	}
+
+	async deleteManyByExecutionIds(executionIds: string[]) {
+		await this.getManager(this.mode).deleteManyByExecutionIds(executionIds);
+	}
+
+	async duplicateBinaryData(inputData: Array<INodeExecutionData[] | null>, executionId: string) {
```

Comment:
We have a lot of code around duplicating binary data, that's only triggered via `IExecuteFunctions['executeWorkflow']`, and that method doesn't seem to be called from anywhere. Maybe this is dead-code and should be removed.

But if this is not dead code, I don't think we should be cloning this data in filesystem (or object store). Binary data is is never edited, and we can pass a readonly reference to binary data wherever we expect a copy.

Reply:
Certainly there is a lot more tech debt to address (small listing in the description), but if this bit is not blocking the S3 implementation, then let's tackle this later.

Implementation-wise, if we return a read-only reference instead of cloning the binary data, if we delete an execution we'll delete its binary data, which means that the non-deleted execution will be having a dangling pointer to deleted binary data.